cpu/sam0_common/periph/freqm.c

```diff
@@ -0,0 +1,263 @@
+
+    return 0;
+}
+
+static void _freqm_clock_init(uint8_t pin, uint8_t gclk_src)
+{
+    /* Selection of the Generator and write Lock for FREQM_MSR */
+    GCLK->PCHCTRL[FREQM_GCLK_ID_MSR].reg = GCLK_PCHCTRL_GEN(pin) | GCLK_PCHCTRL_CHEN_Msk;
+    /* Wait for synchronization */
+    while ((GCLK->PCHCTRL[FREQM_GCLK_ID_MSR].reg & GCLK_PCHCTRL_CHEN_Msk) !=
+           GCLK_PCHCTRL_CHEN_Msk) {}
+
+    /* Selection of the Generator and write Lock for FREQM_REF */
+    GCLK->PCHCTRL[FREQM_GCLK_ID_REF].reg = GCLK_PCHCTRL_GEN(gclk_src) | GCLK_PCHCTRL_CHEN_Msk;
+    /* Wait for synchronization */
+    while ((GCLK->PCHCTRL[FREQM_GCLK_ID_REF].reg & GCLK_PCHCTRL_CHEN_Msk) !=
+           GCLK_PCHCTRL_CHEN_Msk) {}
+}
+
+static struct {
+    freqm_callback_t callback;
+    void *context;
+    freqm_t idx;
+} freqm_obj;
+
+static mutex_t sync_lock = MUTEX_INIT_LOCKED;
+
+static void _freqm_enable(uint8_t refnum)
+{
+    FREQM->CFGA.reg = (uint16_t)(FREQM_CFGA_REFNUM(refnum));
+
+    /* Enable DONE Interrupt */
+    FREQM->INTENSET.reg = (uint8_t)FREQM_INTENSET_DONE_Msk;
+
+    /* Enable FREQM */
+    FREQM->CTRLA.reg = (uint8_t)FREQM_CTRLA_ENABLE_Msk;
+
+    /* Wait for Sync */
+    while ((FREQM->SYNCBUSY.reg) != 0U) {}
+}
+
+static void _freqm_disable(void)
+{
+    /* Disable DONE Interrupt */
+    FREQM->INTENCLR.reg = (uint8_t)FREQM_INTENCLR_MASK;
+    /* Disable FREQM */
+    FREQM->CTRLA.reg &= (uint8_t) ~FREQM_CTRLA_ENABLE_Msk;
+    /* Wait for Sync */
+    while ((FREQM->SYNCBUSY.reg) != 0U) {}
+}
+
+static bool _freqm_is_busy(void)
+{
+    bool freqm_is_busy = false;
+
+    if ((FREQM->STATUS.reg & FREQM_STATUS_BUSY_Msk) == FREQM_STATUS_BUSY_Msk) {
+        freqm_is_busy = true;
+    }
+
+    return freqm_is_busy;
+}
+
+int freqm_frequency_get(freqm_t idx, uint32_t *result)
+{
+    uint64_t result_tmp = 0U;
+    int overflow_condition = 0;
+    const struct freqm_config *config = &config_freqm[idx];
+
+    /* Deregister callback */
+    freqm_obj.callback = NULL;
+    freqm_obj.context = NULL;
+
+    _freqm_enable(config->ref_clk_cnt);
+
+    /* Check if measurement is already in progress */
+    if (!_freqm_is_busy()) {
+        /* Clear the Done Interrupt flag */
+        FREQM->INTFLAG.reg = (uint8_t)FREQM_INTFLAG_DONE_Msk;
+
+        /* Start measurement */
+        FREQM->CTRLB.reg = (uint8_t)FREQM_CTRLB_START_Msk;
+    }
+
+    /* Block until measurement is done */
+    mutex_lock(&sync_lock);
+
+    /* Calculate measured frequency */
+    result_tmp = FREQM->VALUE.reg * (uint64_t)(sam0_gclk_freq(config->gclk_src));
+    result_tmp = result_tmp / config->ref_clk_cnt;
+
+    /* Return measured frequency */
+    *result = (uint32_t)result_tmp;
+
+    /* Read overflow status */
+    overflow_condition = ((int)FREQM->STATUS.reg & FREQM_STATUS_OVF_Msk);
+
+    /* Clear overflow status */
+    FREQM->STATUS.reg = (uint8_t)FREQM_STATUS_OVF_Msk;
+
+    _freqm_disable();
+
+    return overflow_condition;
```

Comment:
Should this return a negative value on detecting an overflow?
Further should something like `-ERANGE`or `-EOVERFLOW` be returned?

Reply:
Both `ERANGE` - *Numerical result out of range* and `EOVERFLOW` - *Value too large for defined data type* would work, I think `EOVERFLOW` fits better as the former sounds more like it refers to an input parameter.